### ISSUE TITLE
[Common Lisp] Implement new Concept Exercise: enumeration

### ISSUE THREAD

**Issue by verdammelt** [exercism/v3#1346]
This issue describes how to implement the `enumeration` concept exercise for the Common Lisp track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student about how to do [enumeration] in Common Lisp. To keep things simple this exercise will use only require a few of `loop`'s features (although they could also be done with members of the `do` "family").

## Learning objectives

- Know the existence of `loop`.
- Know how to use "for-as-arithmetic" clause
- Know how to use "for-as-in-list" clause
- Know how to use "for-as-across" clause
- Know how to use `collect` "list-accumulation" clause
- Know how to use `summing` "numeric-accumulation" clause
- Know how to use `if` and `else` "conditional" clause.

## Out of scope

The following will be out of scope for this exercise - but should be mentioned in the `after.md`.

- Other parts of `loop`.
- The `do` "family" of functions.

## Concepts

- enumeration

## Prerequisites

Student will not be taught these topics but will need some familiarity with them to solve the problems. The introduction, instructions and stub code could be written to lower the need of these prerequisites.

- arithmetic
- strings
- ??? more depending upon the story of the exercise.

## Resources to refer to

- [Loop for Black Belts][loop-for-black-belts]
- [Loop Facility][hyperspec-loop]

For example:

### Hints

Hints for the user can include specific examples of what they need to do with `loop`.

### After

The `after.md` file should mention/discuss the following:

- Other parts of `loop`. No need to be exhaustive but perhaps should include `into`, `finally`, `return`, other "for-as" clauses
- 

## Representer

The representer will need to be able to normalize the `loop` macro.

## Analyzer

The analyzer will need to be able to recognize `loop` vs. `do` and give the user some guidance to use `loop` for this exercise. This guidance should not stop the solution from being accepted.

## Implementing

The intention of focusing on `loop` for this exercise is to keep things simple and reduce prerequisites. `loop` hides things like assignment needed for incrementing and collection than `do` does.

Compare:

    `(loop for i from 0 upto 10 collect i)` 

vs.

    `(do ((i 0 (1+ i)) (result (list) (cons i result))) ((> i 10) (nreverse result)))`

### Suggestion for exercise tasks

- The student must sum even integers in a range. (for-as-arithmetic)
- The student must square all numbers in a list. (for-as-in-list)
- The student must find every vowel in a string. (for-as-across)

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

[enumeration]: ./reference/concepts/enumeration.md
[hyperspec-loop]: http://www.lispworks.com/documentation/HyperSpec/Body/06_a.htm
[loop-for-black-belts]: http://www.gigamonkeys.com/book/loop-for-black-belts.html

**Comment by TheLostLambda** [comment]
I think this looks pretty good overall. Will we teach `do` in some other exercise? Additionally, I had concepts listed for `loop.basic`, `loop.collecting`, `loop.traversal`, and `loop.ranges` (and a couple of others not covered directly in this exercise). Should we:

1. Replace these `loop.` concepts with the single `enumeration` concept
2. Break up this concept into smaller ones
3 Or have this exercise teach several of these smaller ones

**Comment by verdammelt** [comment]
`do` vs. `loop`. When I sat down to write this issue I thought it was going to say use `do` not `loop`  - but I thought that `do` would be harder for early students. And especially with `dotimes` and `dolist` which evaluate to `nil` so we'd need to have side-effects (printing or assignment) so it just seemed to me that `loop` would be easier. 

Also I think/hope that by introducing only a few clauses of `loop` (for-as-arithmetic, for-as-in-list, for-as-across, collecting) it would still be a small enough exercise.  Sorry I had forgotten about the `loop` breakdown you had done.

So I suggest having this exercise teach a few basic `loop` features, leaving others for later. And perhaps a later concept for `do`. (or not... if we teach `loop` as the enumeration construct we don't need to teach `do` at all, the student can learn it on their own if desired.)

I'm starting to see a pattern where we want to teach X but you can do Y or Z as well and we have no way to enforce the use of X. Just me?

**Comment by TheLostLambda** [comment]
> I'm starting to see a pattern where we want to teach X but you can do Y or Z as well and we have no way to enforce the use of X. Just me?

No, I definitely agree. I suppose that's the point of the analyser... I do think that concepts should have singular "ideal" solutions though. I think we've been abiding by that so far. Your `loop` vs `do` example in the implementation section is a good example of that I think.

> So I suggest having this exercise teach a few basic loop features

That sounds good to me! And that "breakdown" of `loop` is also very WIP, so feel free to change it as you see fit. Maybe we can table `do` for now, or mention it in the `after.md` as an alternative. Mentioning it there would be a good way to make the student aware of it without forcing them to learn it when `loop` can usually act as a substitute.

**Comment by verdammelt** [comment]
I'll fold these comments into the issue text. I had meant to put `do` in the `after` but it looks like I failed...
